src/core/vertexmodel.h

```diff
@@ -0,0 +1,188 @@
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+#ifndef VERTEXMODEL_H
+#define VERTEXMODEL_H
+
+#include <QStandardItemModel>
+
+class MapSettings;
+
+#include "qgspoint.h"
+#include "qgsgeometry.h"
+#include "qgscoordinatetransform.h"
+
+/**
+ * @brief The VertexModel class is a model to highlight and edit vertices.
+ * The model is used in map coordinates.
+ * There are different modes: no editing, edit (move/remove) nodes, add nodes (to be implemented)
+ */
+class VertexModel : public QStandardItemModel
+{
+    Q_OBJECT
+    //! The current mode
+    Q_PROPERTY( EditingMode editingMode READ editingMode WRITE setEditingMode NOTIFY editingModeChanged )
+    //! The current point being edited \see editingMode. The expected CRS to read/write is the map canvas CRS
+    Q_PROPERTY( QgsPoint currentPoint READ currentPoint WRITE setCurrentPoint NOTIFY currentPointChanged )
+    //! Map settings is used to define the map canvas CRS and detect any extent change
+    Q_PROPERTY( MapSettings *mapSettings READ mapSettings WRITE setMapSettings NOTIFY mapSettingsChanged )
+    //! number of points in the model
+    Q_PROPERTY( int vertexCount READ vertexCount NOTIFY vertexCountChanged )
+    //! determines if the model has changes
+    Q_PROPERTY( bool dirty READ dirty NOTIFY dirtyChanged )
+    //! determines if one can remove current vertex
+    Q_PROPERTY( bool canRemoveVertex READ canRemoveVertex NOTIFY canRemoveVertexChanged )
+    //! determines if one can add vertex
+    Q_PROPERTY( bool canAddVertex READ canAddVertex NOTIFY canAddVertexChanged )
+    //! determines if one can go to previous vertex
+    Q_PROPERTY( bool canPreviousVertex READ canPreviousVertex NOTIFY canPreviousVertexChanged )
+    //! determines if one can go to next vertex
+    Q_PROPERTY( bool canNextVertex READ canNextVertex NOTIFY canNextVertexChanged )
+
+  public:
+    enum ColumnRole
+    {
+      PointRole = Qt::UserRole + 1,
+      CurrentVertexRole,
+      SegmentVertexRole,
+    };
+
+    enum EditingMode
+    {
+      NoEditing,
+      EditVertex,
+      AddVertex
+    };
+    Q_ENUM( EditingMode )
+
+
+    struct Centroid
+    {
+      QgsPoint point;
+      int index;
+    };
+
+    explicit VertexModel( QObject *parent = nullptr );
+    ~VertexModel() override = default;
+
+    //! \copydoc mapSettings
+    void setMapSettings( MapSettings *mapSettings );
+    //! \copydoc mapSettings
+    MapSettings *mapSettings();
+
+    Q_INVOKABLE void setGeometry( const QgsGeometry &geometry, const QgsCoordinateReferenceSystem &crs );
```

Comment:
Should the CRS be a property? It somehow feels weird to have the geometry setter different to the getter.

Reply:
what I did for something similar recently was this
https://github.com/opengisch/QField/pull/349/files#diff-42599fbe3ce295020d476f85594b8868R125

it basically would not reconstruct the geometry immediately, but only in the next event loop run. If multiple properties are set at once, that will only trigger one expensive recalculation.

But I'd be happy with a simple getter as you propose too.